### PR TITLE
InstreamVideo Implementation

### PR DESCRIPTION
**Pull request by ppuviarasu**

### CODE REVIEW

File: Source/AdUnits/InStreamVideoAdUnit.swift [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+public class InStreamVideoAdUnit: VideoBaseAdUnit {
+    
+    public init(configId: String) {
+        super.init(configId: configId, size: CGSize(width: UIScreen.main.bounds.width, height: UIScreen.main.bounds.height))
+    }
+    
+    //TODO: dynamic is used by tests
+    
+    dynamic public func fetchInstreamDemandForAdObject(adUnitId:String, completion: @escaping(_ result: ResultCode, _ adServerURL:String?) -> Void) {
```
Comment (anwzhang) [inline Source/AdUnits/InStreamVideoAdUnit.swift:26]: This file is unnecessary.
Can you remove it and also update the demo app?

File: Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift [hunk 2]
```diff
@@ -0,0 +1,154 @@
+//
+//  InstreamVideoViewController.swift
+//  PrebidDemoSwift
+//
+//  Created by Punnaghai Puviarasu on 9/28/20.
+//  Copyright © 2020 Prebid. All rights reserved.
+//
```
Comment (yoalex5) [inline Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift:7]: can we change it ?

File: Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+//    + "output=vast&unviewed_position_start=1&"
+//    + "cust_params=sample_ct%3Dlinear%26hb_uuid_appnexus%3Dd2913b6a-6abb-4451-941c-12657097c5bc%26hb_cache_host%3Dprebid.nym2.adnxs.com&hb_bidder_appnexus%3Dappnexus%26hb_size_appnexus%3D1x1%26hb_pb_appnexus%3D0.50%26hb_cache_path%3D/pbc/v1/cache%26hb_pb%3D0.50%26hb_cache_path_appnex%3D/pbc/v1/cache%26hb_uuid%3Dd2913b6a-6abb-4451-941c-12657097c5bc%26hb_size%3D1x1%26hb_env%3Dmobile-app%26hb_env_appnexus%3Dmobile-app%26hb_bidder%3Dappnexus%26hb_cache_host_appnex%3Dprebid.nym2.adnxs.com"
+    
+    //env
+    //gdfp_req
+    //iu
+    //output
+    //sz - Size of master video ad slot.
+    //unviewed_position_start - Setting this to 1 turns on delayed impressions for video.
+    //cust_params
```
Comment (yoalex5) [inline Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift:38]: can we remove it ?

File: Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    
+    static let kTestAppAdTagUrl =
+    "https://pubads.g.doubleclick.net/gampad/ads?sz=640x480&"
+    + "iu=/19968336/Punnaghai_Instream_Video1&gdfp_req=1&env=vp&"
+    + "output=vast&unviewed_position_start=1&"
+    + "cust_params=hb_size_appnexus%3D1x1%26hb_cache_id_appnexus%3D37a2083f-f7ab-4f75-9c68-ef6e65de60bb%26hb_uuid%3Dda6b9303-52fc-434b-96e1-786c7f006735%26hb_env_appnexus%3Dmobile-app%26hb_cache_host%3Dprebid.nym2.adnxs.com%26hb_cache_path%3D%2Fpbc%2Fv1%2Fcache%26hb_env%3Dmobile-app%26hb_size%3D1x1%26hb_cache_host_appnex%3Dprebid.nym2.adnxs.com%26hb_cache_id%3D37a2083f-f7ab-4f75-9c68-ef6e65de60bb%26hb_bidder_appnexus%3Dappnexus%26hb_bidder%3Dappnexus%26hb_pb_appnexus%3D0.50%26hb_cache_path_appnex%3D%2Fpbc%2Fv1%2Fcache%26hb_pb%3D0.50%26hb_uuid_appnexus%3Dda6b9303-52fc-434b-96e1-786c7f006735"
```
Comment (yoalex5) [inline Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift:44]: It is hardcoded

File: Tests/FetchingLogictests/UtilsTests.swift [hunk 5]
```diff
@@ -551,4 +551,49 @@ class UtilsTests: XCTestCase {
...
+//        let extractedKeywords:[String] = newString.components(separatedBy: "&")
+//        
+//        var extractedPrebidKeywords: [String: String] = [String:String]()
+//        
+//        for keyString in extractedKeywords {
+//            let keywords:[String] = newString.components(separatedBy: ":")
+//            
+//            extractedKeywords[keywords[0]] = keywords[1]
+//            
+//        }
```
Comment (yoalex5) [inline Tests/FetchingLogictests/UtilsTests.swift:594]: can we remove it ?

**Review by yoalex5 — COMMENTED** [review]
1. We should add Rubicon In-stream 
2. Please fix CI error

File: Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift (hunk 2 above)
Comment (ppuviarasu) [reply Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift]: done

File: Source/Addendum/IMAUtils.swift [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+public final class IMAUtils: NSObject {
+    
+    @objc
+    public static let shared = IMAUtils()
+    
+    private override init() {}
+    
+    @objc
+    public func constructAdTagURLForIMAWithPrebidKeys (adUnitID:String, customKeywords: [String:String]) -> String{
+        let adServerURL = "https://pubads.g.doubleclick.net/gampad/ads?sz=640x480&output=vast&unviewed_position_start=1&gdfp_req=1&env=vp"
```
Comment (anwzhang) [inline Source/Addendum/IMAUtils.swift:20]: should not be fixed sz
either have the publisher pass in the size or use hb_size

Comment (anwzhang) [reply Source/Addendum/IMAUtils.swift:20]: also android has changed to output=xml_vast4 based on @yoalex5 feedback

File: Source/AdUnits/InStreamVideoAdUnit.swift (hunk 1 above)
Comment (ppuviarasu) [reply Source/AdUnits/InStreamVideoAdUnit.swift]: done

File: Tests/FetchingLogictests/UtilsTests.swift (hunk 5 above)
Comment (ppuviarasu) [reply Tests/FetchingLogictests/UtilsTests.swift]: done

File: Example/PrebidDemo/PrebidDemoSwift/AppDelegate.swift [hunk 7]
```diff
@@ -55,6 +55,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         if #available(iOS 14, *) {
             ATTrackingManager.requestTrackingAuthorization(completionHandler: { status in
             })
+
```
Comment (anwzhang) [inline Example/PrebidDemo/PrebidDemoSwift/AppDelegate.swift:58]: remove this.

File: Example/PrebidDemo/PrebidDemoSwift/BannerController.swift [hunk 8]
```diff
@@ -174,6 +174,7 @@ class BannerController: UIViewController, GADBannerViewDelegate, MPAdViewDelegat
 
         adUnit.fetchDemand(adObject: self.amRequest) { [weak self] (resultCode: ResultCode) in
             print("Prebid demand fetch for AdManager \(resultCode.name())")
+            print(self?.amRequest.contentURL!);
```
Comment (anwzhang) [inline Example/PrebidDemo/PrebidDemoSwift/BannerController.swift:177]: remove this.

File: Source/Addendum/IMAUtils.swift [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    public static let shared = IMAUtils()
+    
+    private override init() {}
+    
+    public func constructAdTagURLForIMAWithPrebidKeys (adUnitID:String, adSlotSizes:[IMAAdSlotSize], customKeywords: [String:String]) -> String{
+        let adServerURL = "https://pubads.g.doubleclick.net/gampad/ads?output=xml_vast4&unviewed_position_start=1&gdfp_req=1&env=vp"
+        
+        var adSlotSize = "sz="
+        
+        for adSlot in adSlotSizes {
```
Comment (anwzhang) [inline Source/Addendum/IMAUtils.swift:44]: what if the publisher passes in the empty array or duplicated sizes?

Comment (ppuviarasu) [reply Source/Addendum/IMAUtils.swift:46]: I have handled empty array. Duplicate adServer doesnt throw any error. So left it

**Review by anwzhang — APPROVED** [review]

File: Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    @IBOutlet var adServerLabel: UILabel!
+    
+    @IBOutlet var appInstreamView: UIView!
+    
+    var adServerName: String = ""
+    
+    private var adUnit: VideoAdUnit!
+    
+    var adsLoader: IMAAdsLoader!
+    var adsManager: IMAAdsManager!
```
Comment (yoalex5) [inline Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift:31]: Please `destroy` adsManager object when it is necessary

Comment (yoalex5) [reply Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift:31]: Please check #337 with Rubicon in-stream

**Comment by yoalex5** [comment]
Please check #337 with Rubicon in-stream

File: Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    func setupAndLoadAMInstreamVideo() {
+        setupPBAppNexusInStreamVideo()
+        setupAMAppNexusInstreamVideo()
+        
+    }
+    
+    //Setup PB
+    func setupPBAppNexusInStreamVideo() {
+
+        setupPB(host: .Appnexus, accountId: "aecd6ef7-b992-4e99-9bb8-65e2d984e1dd", storedResponse: "sample_video_response")
```
[the remark this answers is not in the document]
Comment (ppuviarasu) [reply Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift]: updated from rubicon sample

File: Source/Addendum/IMAUtils.swift [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+@objc public enum IMAAdSlotSize: Int {
+    
+    case Size400x300
+    
+    case Size640x480
+    
+    func size () -> String {
+        switch self {
+        case .Size400x300: return "400x300"
+        case .Size640x480: return "640x480"
```
Comment (yoalex5) [inline Source/Addendum/IMAUtils.swift:27]: why there are only two sizes ?

File: Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import UIKit
+import GoogleInteractiveMediaAds
+import PrebidMobile
+
+class InstreamVideoViewController: UIViewController, IMAAdsLoaderDelegate, IMAAdsManagerDelegate {
```
Comment (yoalex5) [inline Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift:20]: Can we add a video player with a real video

Comment (bszekely1) [reply Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift:20]: This is preferred to showcase to real in-stream integration for publishers to follow. Either present the video ad before the video content or in the middle of content, passing the appropriate signal such as pre-roll or mid-roll or post-roll.

File: Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift (hunk 10 above)
Comment (ppuviarasu) [reply Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift:32]: done

File: Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift (hunk 13 above)
Comment (ppuviarasu) [reply Example/PrebidDemo/PrebidDemoSwift/InstreamVideoViewController.swift:20]: Updated with the content play logic in sample app

File: Source/Addendum/IMAUtils.swift (hunk 12 above)
Comment (ppuviarasu) [reply Source/Addendum/IMAUtils.swift:27]: these are the only 2 sizes IMA supports

Comment (yoalex5) [reply Source/Addendum/IMAUtils.swift:27]: do you have a doc link ?

Comment (ppuviarasu) [reply Source/Addendum/IMAUtils.swift:27]: This is in creating vast creatives in the adsManager. The same implementation can be found in android too...
Screen Shot 2020-11-30 at 8 03 06 PM

**Comment by yoalex5** [comment]
@ppuviarasu 

what do you think about mid-roll and post-roll ads ?

**Comment by ppuviarasu** [comment]
The fetching & playing ads sample remains the same... We pass in video params (mid roll or post roll) in our request if we need to specially mention this to our demand partners. But for the integration the publisher needs to pause the content & play where ever he wants.

**Comment by yoalex5** [comment]
Should we use this sample into out Demo app ?

**Comment by ppuviarasu** [comment]
I dont think it is necessary... I can ask @littleV to chime in...

**Comment by anwzhang** [comment]
> Should we use this sample into out Demo app ?

I don't think we need that sample into the demo app. Since the fetching ads part is exactly the same.

**Review by yoalex5 — APPROVED** [review]

**Pull request merged**